sdhc: esp32s3: Add support to SDHC cards

raffarost opened the pull request:
Added support to SDHC driver for ESP32S3

Edit: PR was updated with commit for `esp32s3_devkitm` SDHC support removed, as it doesn't have an SD slot. A sample will be added to `esp32s3-eye` instead, by a separate PR.

EricNRS commented:
LGTM, I tested and verified this patch on ESP32-S3 hardware in both 1-bit and 4-bit SDIO mode.

danieldegrasse commented on drivers/sdhc/sdhc_esp32.c:
The commit message for these changes needs to be *much* more descriptive- truthfully I'd prefer if these changes were split into a few commits:
- commit adding code to enable SDHC clock on boot
- commit changing handling of the card width fields (the cleanup of the LOG_DBG message can also go here)
- commit for pin control and DTS changes

Ideally each commit (and its message) are sufficient to figure out *why* the change was made. Right now, I can guess that these changes are all needed to enable the ESP32S3 with the SDHC driver. However, it isn't clear to me why the driver changes are needed- which is why I would like them to be in separate commits with more descriptive messages. A commit message like the one you have is fine if all you are doing is adding devicetree and pin control definitions, since those logically only affect that device.

raffarost replied:
hi @danieldegrasse, the changes in the driver are related to S3 in the following way:

- sdmmc_ll_enable_bus_clock() function in S3 would directly access another subsystem (clock), so it was better to make the correction and configure clock using clock_control api instead
- S3 required specific registers to be accessed as 32-bit words, so we included a flag in HAL to enforce 32-bit access to peripheral registers and this allowed for the macros cleanup

The CI errors unfortunately are not related to this PR, so we will fix them soon.

danieldegrasse reviewed (commented):
One comment from my side- please look at CI issues when you have a chance as well

Commit:
dts: pinctrl: esp32s3: Support for SDHC

Added pinctrl and dts settings for SDHC

Signed-off-by: Raffael Rostagno <raffael.rostagno@espressif.com>

Commit:
sdhc: esp32: Clock control setting from DTS

Enable SDIO clock using clock control subsystem instead of
low level functions.

Signed-off-by: Raffael Rostagno <raffael.rostagno@espressif.com>

Commit:
sdhc: esp32: Register writing macros cleanup

Macros to write specific peripheral registers were removed as a
compile flag was added to hal_espressif requesting struct bitfields
to have strict volatile access (thus 32-bit access), allowing for
the cleanup.

Signed-off-by: Raffael Rostagno <raffael.rostagno@espressif.com>

uLipe reviewed (changes requested):
LGTM so far, the only point regarding the commit messages, is the devkitm sdhc support.

It should state only the esp32s3_devkitm board has got the support.

Something like:

`boards: esp32s3_devkitm:  add support for the SDHC ...`

The current message implies that all the s3 boards have gotten support for the SDHC peripheral.

sylvioalves commented on drivers/sdhc/sdhc_esp32.c:
```diff
@@ -1195,7 +1187,7 @@ static int sdhc_esp32_request(const struct device *dev, struct sdhc_command *cmd
 
 	if ((ret_esp != 0) || esp_cmd.error) {
-		LOG_DBG("\nError for command: %u arg %08x ret_esp = 0x%x error = 0x%x\n",
+		LOG_DBG("Error command: %u arg %08x ret_esp = 0x%x error = 0x%x\n",
```
Check compliance.

raffarost replied:
Is this a legit warning? Running locally or checking the Summary we see no errors.
